File: windows/QMK Toolbox/Usb/Bootloader/Ht32DfuDevice.cs
using System.IO;
using System.Threading.Tasks;

namespace QMK_Toolbox.Usb.Bootloader
{
    class Ht32DfuDevice : BootloaderDevice
    {
        public Ht32DfuDevice(UsbDevice d) : base(d)
        {
            Type = BootloaderType.Ht32Dfu;
            Name = "HT32 DFU";
            PreferredDriver = "WinUSB";
            IsResettable = true;
        }

        public async override Task Flash(string mcu, string file)
        {
            if (Path.GetExtension(file)?.ToLower() == ".bin")
            {
                await RunProcessAsync("ht32-dfu-tool.exe", $"--reset write 0 \"{file}\"");
            }
            else
            {
                PrintMessage("Only firmware files in .bin format can be flashed with ht32-dfu-tool!", MessageType.Error);
            }
        }

        public async override Task Reset(string mcu) => await RunProcessAsync("ht32-dfu-tool.exe", "reset");
    }
}

File: macos/QMK Toolbox/USB/Bootloader/HT32DFUDevice.swift
import Foundation

class HT32DFUDevice: BootloaderDevice {
    override init(usbDevice: USBDevice) {
        super.init(usbDevice: usbDevice)
        name = "HT32 DFU"
        type = .ht32Dfu
        resettable = true
    }

    override func flash(_ mcu: String, file: String) {
        guard file.lowercased().hasSuffix(".bin") else {
            print(message: "Only firmware files in .bin format can be flashed with ht32-dfu-tool!", type: .error)
            return
        }

        runProcess("ht32-dfu-tool", args: ["--reset", "write", "0", file])
    }

    override func reset(_ mcu: String) {
        runProcess("ht32-dfu-tool", args: ["reset"])
    }
}
